migrations: Add migration fixing realm of Attachments.

mateuszmandera opened the pull request:
Old messages with attachments, sent by cross-realm bots, didn't have
the realm set correctly to the target realm, this migration fixes it
by looking at the messages linking to the attachment and setting the realm
based on the recipient.

This is the migration follow-up to
410281624098f56ff9a67c8c17aca205487d59bf

mateuszmandera commented on zerver/migrations/0386_fix_attachment_realm.py:
```diff
@@ -0,0 +1,65 @@
+    internal_realm = Realm.objects.get(string_id=settings.SYSTEM_BOT_REALM)
+
+    print("")
+    for attachment in Attachment.objects.filter(realm_id=internal_realm.id).order_by("id"):
+        print(f"Fixing attachment {attachment.id}")
+        message = attachment.messages.only("recipient").first()
+        if message is None:
+            # If attachment.messages is empty, nothing to do.
+            continue
+
+        message_realm_id = get_realm_id_from_recipient(message.recipient)
+        attachment.realm_id = message_realm_id
+        attachment.save(update_fields=["realm_id"])
```
Do we want to re-write this logic for proper bulk operations? I'm not certain if it's worth it though, because that's probably only needed for Zulip Cloud and a simpler migration strategy there might be to just add some `time.sleep()` here and let the code slowly run in the background? (could even just be done by copy-pasting into `manage.py shell` rather than using actually `manage.py migrate`)

alexmv replied:
In Zulip Cloud, there are 27k messages that this touches.  The rows in `zerver_attachment` aren't that wide, so this is probably also OK from a memory point of view.

timabbott replied:
Yeah, there's no need to do bulk operations here, since it's a relatively rare situation that created rows needing this attention.

timabbott commented:
This probably needs to be adjusted to handle ArchiveAttachment objects as well (I verified those exist in Zulip Cloud). Otherwise lgtm!

timabbott commented:
Just a note that we might be able to rewrite this more simply to use the new `message.realm` field.

alexmv commented on zerver/migrations/0423_fix_attachment_realm.py:
```diff
@@ -0,0 +1,50 @@
+        for attachment in model.objects.filter(realm_id=internal_realm.id).order_by("id"):
+            print(f"Fixing {model.__name__} {attachment.id}")
+            message = attachment.messages.only("realm_id").first()
+            if message is None:
+                # If attachment.messages is empty, nothing to do.
+                continue
```
For attachments that #24128 fixes, they don't have any messages on them.  However, the S3 backend _did_ use `target_realm.id` and not `user_profile.realm_id` when calling `self.generate_message_upload_path` so the `path_ids` start with the realm they _should_ be in.

So I think we can also try matching `path_id` against `^(\d+)/` (because there are some junk path-ids which start with `s3.amazonaws.com/`) and use that value as the proposed `realm_id` if it's different than the attachment's (it may be the same if the server was using local-file storage).

mateuszmandera replied:
@alexmv Hmm thinking about this some more - the old bug that this migration is addressing was different than https://github.com/zulip/zulip/pull/24128 in that there was no failure to attach the file to messages resulting from the bug. So `attachment.messages` would only be empty if there truly are no messages linking to it. And at that point, its `realm_id` doesn't matter as it is "unclaimed" and should be actually getting deleted?

For `Attachment` that would overlap the 410281624098f56ff9a67c8c17aca205487d59bf bug and the #24128 bug (not sure if that's possible?), your migration from #24128 would have already fixed `attachment.messages`, so those are not relevant here either.

So bottom line - I'd suggest it's not worth complicating the migration's logic for these `Attachment`s since they are truly unclaimed and qualified for deletions?

alexmv replied:
The problem is that #24128 _didn't_ fix attachments subject to the 410281624098f56ff9a67c8c17aca205487d59bf bug, since that migration looked for messages in the same realm as the attachment when looking for messages to attach the attachment to:
https://github.com/zulip/zulip/blob/30204ed694f55e5556ac39796588b33c3cfdb1b5/zerver/migrations/0423_fix_email_gateway_attachment_owner.py#L62-L69

But for old attachments, those have a realm of the system realm, meaning that it didn't find any matching messages.

Unfortunately, this is now moot for Zulip Cloud and CZO because we've run `delete_old_unclaimed_attachments`, which nuked them all. :(

There may exist other old installs which have not yet upgraded to having a `delete_old_unclaimed_attachments` cron, however.  I think we have to options:
1. Update this migration to, when it finds orphan attachments, parse the potential realm_id out of the path_id, and move them accordingly.  This will leave them orphaned but in the right realm, so we'll need a repeat of the migration in #24128, which can make its previous migration into a no-op.  This has the advantage of fixing anyone who ran the #24128 migration but somehow hasn't run `delete_old_unclaimed_attachments` yet.
2. Change the currently-merged #24128 migration to parse the realm_id out of the path_id if `attachment.realm_id` is the system realm, and leave this migration as-is.  For folks that haven't run it yet, that will reattach these system-realm-attachments to their proper places, and this migration move things to the right realm.  This has the advantage of being slightly less futzy, though it won't quite fix attachments which were not fixed by #24128 when it was first merged if they survived a `delete_old_unclaimed_attachments` culling.

mateuszmandera replied:
@alexmv Oh yeah, I see now. I went with updating this migration here to infer from `path_id`s, so option (1). Can you review the logic?

Also I still need to add a commit to re-run #24128 Is the right way to do it to purge `zerver/migrations/0423_fix_email_gateway_attachment_owner.py` to be a noop and instead move the logic into a new migration? I seem to recall you did some kind of migration repeat like that recently but I can't seem to remember which one to look up the commit

alexmv replied:
35e27aef4a3c9700753149eb71768fc313889d8d, maybe?

But yes, making the previous migration a no-op and putting a fresh copy on the tip of the migration tree is the right way to do this, I think.

mateuszmandera commented on zerver/migrations/0435_fix_attachment_realm.py:
```diff
@@ -0,0 +1,68 @@
+                try:
+                    realm_id = int(matches[0])
+                except ValueError:
+                    # Don't do anything if path_id doesn't start with a sensible realm id.
```
Hmm yeah, if `Attachment`s like that exist and didn't get deleted then this would fail the migration. Tweaked with this to handle the edge case:
```
                    if not Realm.objects.filter(id=realm_id).exists():
                        # If the realm doesn't exist (e.g. due to deletion), we can't do anything.
                        continue
```

mateuszmandera commented:
@alexmv  Fixed https://github.com/zulip/zulip/pull/21580#discussion_r1168912418 and added changes to noop `zerver/migrations/0423_fix_email_gateway_attachment_owner.py` and re-run in `0439`

alexmv commented:
This looks to have never been merged, and I'm not sure whether anything equivalent was either.

mateuszmandera commented on zerver/migrations/0444_rerun_fix_email_gateway_attachment_owner.py:
```diff
@@ -0,0 +1,179 @@
+    bot.  Find each such orphaned attachment, and re-own it and attach
+    it to the appropriate message.
+
+    This was previously zerver/migrations/0443_fix_attachment_realm.py,
```
note to self: wrong migration referenced

mateuszmandera commented:
Rebased the migration.

There was also a second part to this, per https://github.com/zulip/zulip/pull/21580#discussion_r1157927818, we wanted  to re-run the migration from #24128.
But that hinged on the idea that most servers didn't run the `delete_old_unclaimed_attachments` job which would have  deleted the attachments we aimed to fix. 

We added the cronjob in 6b37f9a290a93207c382cfd0fc8b7e41a4152955 in 7.0. Given that's about to be 3 release cycles behind, do we still want to re-run even though most servers probably won't benefit unless they've very old? @alexmv What do you think?

timabbott commented on zerver/migrations/0642_fix_attachment_realm.py:
```diff
@@ -0,0 +1,71 @@
+                        continue
+                except ValueError:
+                    # Don't do anything if path_id doesn't start with a sensible realm id.
+                    continue
```
Should either of these situations log something?

mateuszmandera replied:
I think this one is a pretty normal scenario?
```
                    if not Realm.objects.filter(id=realm_id).exists():
                        # If the realm doesn't exist (e.g. due to deletion), we can't do anything.
                        continue
```

But this would be strange, (not sure if even possible given that we must have gotten a regex match `^(\d+)\/`?

So actually not finding a match is more interesting. Added logging for it here:
```
                matches = re.findall(r"^(\d+)\/", attachment.path_id)
                if not matches:
                    print(f"No realm_id found in path_id '{attachment.path_id}' of attachment {attachment.id}")
                    continue
```

timabbott commented on zerver/migrations/0642_fix_attachment_realm.py:
```diff
@@ -0,0 +1,71 @@
+                    continue
+
+            attachment.realm_id = realm_id
+            attachment.save(update_fields=["realm_id"])
```
This should check if `attachment.realm_id == realm_id`, and continue if so. That should routinely be the case in those organizations where the system bots are in a normal realm, and it's nice to skip database work.

timabbott commented on zerver/migrations/0642_fix_attachment_realm.py:
```diff
@@ -0,0 +1,71 @@
+    print()
+    for model in [Attachment, ArchivedAttachment]:
+        for attachment in model.objects.filter(realm_id=internal_realm.id).order_by("id"):
+            print(f"Fixing {model.__name__} {attachment.id}")
```
Let's move this print statement after we return early if the computed realm is the internal realm.

mateuszmandera commented:
Updated.

timabbott commented:
Added this to make the logging a bit nicer:
``` diff
diff --git a/zerver/migrations/0672_fix_attachment_realm.py b/zerver/migrations/0672_fix_attachment_realm.py
index 658b44ac78..56a6633664 100644
--- a/zerver/migrations/0672_fix_attachment_realm.py
+++ b/zerver/migrations/0672_fix_attachment_realm.py
@@ -53,7 +53,9 @@ def fix_attachment_realm(apps: StateApps, schema_editor: BaseDatabaseSchemaEdito
                 # It's already correct, nothing to do.
                 continue
 
-            print(f"Fixing {model.__name__} {attachment.id}")
+            print(
+                f"Fixing incorrect realm for {model.__name__} {attachment.id} ({attachment.realm_id} => {realm_id})"
+            )
             attachment.realm_id = realm_id
             attachment.save(update_fields=["realm_id"])
```
